expose SQL Obfuscation to python checks

djova opened the pull request:
### What does this PR do?

Exposes sql obfuscation from `pkg/trace/obfuscate` to python checks:
  * required methods, structs, and fields were made public to allow for doing sql obfuscation directly on a string

### Motivation

Goal is to enable any python check dealing with raw sql queries to obfuscate and normalize them the same way we do for traces. 

### Describe your test plan

Unit tests cover expected obfuscation feature as well as expected response to a variety of invalid input error cases.

hush-hush commented on pkg/collector/python/util.go:
```diff
@@ -79,3 +81,15 @@ func GetSubprocessOutput(argv **C.char, cStdout **C.char, cStderr **C.char, cRet
+	s := C.GoString(rawQuery)
+	obfuscatedQuery, err := obfuscator.ObfuscateSQLString(s)
+	if err != nil {
+		return TrackedCString(fmt.Sprint("ERROR: %v", err))
```
We should return `nil` here and check the value in C. That way we can set the correct error flag in python

djova replied:
Updated

hush-hush commented on pkg/collector/python/util.go:
```diff
@@ -79,3 +81,15 @@ func GetSubprocessOutput(argv **C.char, cStdout **C.char, cStderr **C.char, cRet
 	*cRetCode = C.int(retCode)
 }
+
+var obfuscator = obfuscate.NewObfuscator(nil)
```
Is there a way not to have this in global ? Is it costly to instantiate every time ? Does the obfucator keep an internal state ?

djova replied:
It doesn't keep state. It's pretty lightweight, mostly [this config struct](https://github.com/DataDog/datadog-agent/blob/983b9792704458a4a6c7aa820a4a7385ede1765a/pkg/trace/obfuscate/obfuscate.go#L26-L25), which with the default settings is ~ 25 bytes. 

It'll need to be re-allocated on the heap each time the function runs if we recreate the obfuscator. Probably not a big deal given that this will be called only from python.

Is there a reason to avoid keeping it as a global?

hush-hush replied:
We try to allow package to be tear down or created multiple times if needed (for example you can instantiate multiple forwarder or aggregator if needed). The design of the agent was to have flexible package in `pkg` put together by binaries in `cmd`.

It it doesn't keep state I think this is fine

hush-hush commented on rtloader/common/builtins/util.c:
```diff
@@ -8,14 +8,18 @@
 
 static PyMethodDef methods[] = {
     { "headers", (PyCFunction)headers, METH_VARARGS | METH_KEYWORDS, "Get standard set of HTTP headers." },
+    { "obfuscate_sql", (PyCFunction)obfuscate_sql, METH_VARARGS, "Obfuscate Sql." },
```
This should be move to the `datadog_agent` builtin. `util` package is a legacy requirement from A5.

hush-hush commented on rtloader/include/rtloader_types.h:
```diff
@@ -125,6 +125,8 @@ typedef void (*cb_set_external_tags_t)(char *, char *, char **);
 typedef void (*cb_write_persistent_cache_t)(char *, char *);
 // (value)
 typedef char *(*cb_read_persistent_cache_t)(char *);
+// (value)
```
we try to explain what is the parameter here. It's the "value" when doing a write to the cache. The comment for obfuscate_sql could be `// (sql_query)`

djova replied:
Updated

hush-hush commented on rtloader/common/builtins/util.c:
```diff
@@ -52,3 +61,42 @@ PyObject *headers(PyObject *self, PyObject *args, PyObject *kwargs)
+    v = cb_obfuscate_sql(key);
+
+    if (v == NULL) {
+        PyErr_SetString(PyExc_RuntimeError, "failed to read data");
```
Should be `failed to obfuscate sql query`, no ?

djova replied:
Yes updated to contain the obfuscation error.

djova committed:
expose agent sql obfuscation to python checks

Exposes sql obfuscation from `pkg/trace/obfuscate` to python checks:
  * required methods, structs, and fields were made public to allow for doing sql obfuscation directly on a string

Goal is to enable any python check dealing with raw sql queries to obfuscate and normalize them the same way we do for traces.

hush-hush commented on pkg/collector/python/datadog_agent.go:
```diff
@@ -185,3 +187,20 @@ func ReadPersistentCache(key *C.char) *C.char {
+	obfuscatedQuery, err := obfuscator.ObfuscateSQLString(s)
+	if err != nil {
+		// memory will be freed by caller
+		*errResult = TrackedCString(fmt.Sprintf("%v", err))
```
We should use err.Error() instead

djova replied:
Updated

hush-hush commented on rtloader/common/builtins/datadog_agent.c:
```diff
@@ -653,3 +662,46 @@ static PyObject *set_external_tags(PyObject *self, PyObject *args)
+
+    char *rawQuery;
+    if (!PyArg_ParseTuple(args, "s", &rawQuery)) {
+        return NULL;
```
You need to release the GIL here since it was acquired line 684

djova replied:
Oh whoops, good catch! Updated

hush-hush commented on rtloader/common/builtins/datadog_agent.c:
```diff
@@ -653,3 +662,46 @@ static PyObject *set_external_tags(PyObject *self, PyObject *args)
+        PyErr_SetString(PyExc_RuntimeError, err);
+        cgo_free(err);
+    }
+    if (obfQuery != NULL) {
```
We should set an error in case both `err` and and `obfQuery` are null

djova replied:
If they're both null why not just return null?

hush-hush replied:
Because `NULL` is not a PyObject (`PyNone` is the equivalent). But more importantly if both are null we should raise an error as the Go code most likely misbehaved. 

So either PyNone or and error. A user on the python side should not have to check for `None` but catch an exception, that's why I think we should raise one here.

djova replied:
Good point. Updated to set an error in this case. Please take another look!

hush-hush commented on rtloader/common/builtins/datadog_agent.c:
```diff
@@ -653,3 +662,46 @@ static PyObject *set_external_tags(PyObject *self, PyObject *args)
+    }
+
+    char *obfQuery = NULL;
+    char *err = NULL;
```
Could we rename `err` to `error_message` to be explicit ?

djova replied:
Renamed

hush-hush commented on rtloader/common/builtins/datadog_agent.c:
```diff
@@ -653,3 +662,46 @@ static PyObject *set_external_tags(PyObject *self, PyObject *args)
+    obfQuery = cb_obfuscate_sql(rawQuery, &err);
+
+    PyObject *retval = NULL;
+    if (err != NULL) {
+        PyErr_SetString(PyExc_RuntimeError, err);
+        cgo_free(err);
+    }
```
If both `err` and `obfQuery` are not null we will set an error that will be raise the next time a Cpython function return null.
I think if `err` is not null we should raise the exception (free `obfQuery` of needed) and return NULL.

djova replied:
Yes good point! Updated to return null if error, regardless of obfQuery result. Does that work?

hush-hush commented on rtloader/include/rtloader_types.h:
```diff
@@ -125,6 +125,8 @@ typedef void (*cb_set_external_tags_t)(char *, char *, char **);
 typedef void (*cb_write_persistent_cache_t)(char *, char *);
 // (value)
 typedef char *(*cb_read_persistent_cache_t)(char *);
+// (sql_query, error)
```
Could we say `// (sql_query, error_message)` ?

djova replied:
Renamed

hush-hush commented on pkg/collector/python/init.go:
```diff
@@ -121,9 +121,11 @@ void initAggregatorModule(rtloader_t *rtloader) {
 //
 
 void GetSubprocessOutput(char **, int, char **, char **, int*, char **);
+char* ObfuscateSql(char *, char **);
 
 void initUtilModule(rtloader_t *rtloader) {
 	set_get_subprocess_output_cb(rtloader, GetSubprocessOutput);
+	set_obfuscate_sql_cb(rtloader, ObfuscateSql);
```
Those are part of the `datado_agent` module now.

djova replied:
Good catch, moved

hush-hush reviewed (changes requested):
Looks good overall, just a few details

justiniso commented on pkg/collector/python/datadog_agent.go:
```diff
@@ -185,3 +186,20 @@ func ReadPersistentCache(key *C.char) *C.char {
+	obfuscatedQuery, err := obfuscator.ObfuscateSQLString(s)
+	if err != nil {
+		// memory will be freed by caller
+		*errResult = TrackedCString(err.Error())
```
These comments are less of a review and more for my own education, but how does this error actually get passed to the C code and raised in python? It looks like the error is set here: `obfQuery = cb_obfuscate_sql(rawQuery, &error_message);` but there's something magic I'm missing, can you explain?

djova replied:
This error string gets written into `error_message`, then we from c code we tell python to throw a runtime error containing that message using `PyErr_SetString(PyExc_RuntimeError, error_message);` ([more info here](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_SetString))

justiniso replied:
🤦 I completely missed the `errResult` in the signature here. No magic, this makes sense.

hush-hush commented on rtloader/common/builtins/datadog_agent.c:
```diff
@@ -653,3 +662,49 @@ static PyObject *set_external_tags(PyObject *self, PyObject *args)
+        PyErr_SetString(PyExc_RuntimeError, error_message);
+        cgo_free(error_message);
+        // there shouldn't be any return value if there was an error, but null it out just in case
+        retval = NULL;
```
If `obfQuery` wasn't `NULL` we just leaked a PyObject created line 698.

We should first check for error and `cgo_free(obfQuery)` (a NULL pointer will be discarded by cgo_free.
Then if not error, create the retval.

djova replied:
Good catch! Updated

hush-hush commented on rtloader/test/datadog_agent/datadog_agent.go:
```diff
@@ -8,6 +8,7 @@ import (
 	"strings"
 	"unsafe"
 
+	"github.com/DataDog/datadog-agent/pkg/trace/obfuscate"
```
Please don't link rtloader to any other package from the agent. It's not needed and put an hard dependencies between both. Technically rtloader function could be initialized with any go code, so let's leverage this in our test and keep both of them separate.

djova replied:
Ok updated to remove the dependency

hush-hush commented on rtloader/test/datadog_agent/datadog_agent.go:
```diff
@@ -229,3 +232,16 @@ func writePersistentCache(key, value *C.char) {
+//export obfuscateSQL
+func obfuscateSQL(rawQuery *C.char, errResult **C.char) *C.char {
+	s := C.GoString(rawQuery)
+	obfuscatedQuery, err := obfuscate.NewObfuscator(nil).ObfuscateSQLString(s)
```
We can just not call anything here and return/set the correct value depending on the input. The obfuscate code is already tested in it's own module. We're only testing how rtloader handles the returned value here.

djova replied:
Ok, updated the remove the dependency

hush-hush commented on rtloader/common/builtins/datadog_agent.c:
```diff
@@ -116,6 +119,12 @@ void _set_tracemalloc_enabled_cb(cb_tracemalloc_enabled_t cb)
+    cb_obfuscate_sql = cb;
+}
+
+
```
A nit but could you remove the extra new line here.

djova replied:
Removed

hush-hush commented on pkg/collector/python/datadog_agent.go:
```diff
@@ -10,6 +10,7 @@ package python
 import (
 	"unsafe"
 
+	"github.com/DataDog/datadog-agent/pkg/trace/obfuscate"
```
I forgot to ask: is the `obfuscate` works on windows ?

djova replied:
I guess? The trace agent runs on windows and it already runs all sql queries that come through trace data, so it must work to some extent already.

djova replied:
Confirmed that none of the `!windows` build flags in `pkg/trace` have an impact on the obfuscation code, so it should work the same on windows from a build perspective.
